firebase-firestore/src/main/java/com/google/firebase/firestore/local/MemoryDocumentOverlays.java

```diff
@@ -0,0 +1,54 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.local;
+
+import androidx.annotation.Nullable;
+import com.google.firebase.firestore.model.DocumentKey;
+import com.google.firebase.firestore.model.mutation.Mutation;
+import java.util.HashMap;
+import java.util.Map;
+
+public class MemoryDocumentOverlays implements DocumentOverlays {
+  private Map<String, Map<DocumentKey, Mutation>> overlays = new HashMap<>();
+
+  @Nullable
+  @Override
+  public Mutation getOverlayMutation(String uid, DocumentKey key) {
```

Comment:
The way UserIds are handled in the MutationQueue is different. We have a single MutationQueue per uses.  That means that these methods wouldn't need to take the User ID at all.

It also means that you wold not have to implement user switching in this class. Overlays should be destroyed once a user signs out, and not be available once the user signs back in again (to match current Memory-persistence behavior). I noticed that you did not implement User ID switching - so you are already on the right track :)